posts/Martin/2015-07-13-introduction-to-hibernate-search-genericjpa.adoc

```diff
@@ -0,0 +1,460 @@
+org.hibernate.search.genericjpa.db.events.triggers.MySQLTriggerSQLStringSource
+ 
+# we have a searchfactory that is supposed 
+#to get it's updates from SQL
+org.hibernate.search.genericjpa.searchfactory.type=sql
+ 
+# how many updates should be processed at once?
+# this number is quite conservative
+org.hibernate.search.genericjpa.searchfactory.batchSizeForUpdates=10
+ 
+# what delay do we want between checking for updates?
+org.hibernate.search.genericjpa.searchfactory.updateDelay=25
+ 
+# we are in a Java SE environment in this example.
+# if you specify 'true', you will have to add an extra property
+# to lookup the JTA mechanism. This is however not needed
+# if you are managing your Transactions by yourself
+# org.hibernate.search.genericjpa.searchfactory.transactionManagerProvider.jndi=\
+# java:jboss/TransactionManager
+org.hibernate.search.genericjpa.searchfactory.useJTATransactions=false
+ 
+# for this example we store our directory in ram.
+hibernate.search.default.directory_provider=ram
+ 
+# we could add extra properties as specified in the Hibernate Search 
+# documentation (http://docs.jboss.org/hibernate/search/5.3/reference/en-US/html_single/#_configuration) next
+----
+
+Now we could start using hibernate search for queries.
+
+[source,java]
+----
+EntityManager em = ...;
+em.getTransaction().begin();
+Book book = new Book();
+book.setName( "The Hobbit" );
+book.setAuthor( "J.R.R. Tolkien");
+em.persist( book );
+em.getTransaction().commit();
+ 
+FullTextEntityManager fem = searchController.getFullTextEntityManager( em );
+ 
+QueryBuilder qb = fem.getSearchFactory()
+  .buildQueryBuilder().forEntity(Book.class).get();
+org.apache.lucene.search.Query query = qb
+  .keyword()
+  .onFields("name")
+  .matching("The Hobbit")
+  .createQuery();
+ 
+List books = fem.createFullTextQuery(
+  query, Book.class )
+    .getResultList();
+----
+
+But wait, why didn’t we receive any Books even though our query was right?
+Well, that’s because we forgot to add some additional classes to our persistence.xml.
+In order to keep the index up-to-date we have to specify special entities that can hold
+the information about changes in the database. These will be queried by our
+Hibernate Search GenericJPA engine and then the index is updated accordingly.
+For our example, we need this special entity:
+
+[source,java]
+----
+package org.hibernate.search.genericjpa.test.entities;
+ 
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+ 
+import org.hibernate.search.genericjpa.annotations.Event;
+import org.hibernate.search.genericjpa.annotations.IdFor;
+import org.hibernate.search.genericjpa.annotations.Updates;
+ 
+@Entity(name = "BookUpdates")
+// what is the original table name? and what table is this entity saved to
+// this is kinda duplicated, but needed
+@Updates(originalTableName = "Book", tableName = "BookUpdates")
+@Table(name = "BookUpdates")
```

Comment:
> +1 I think there is ways to much overhead involved here. Additional entities (could they btw not be generated!?), update tables, triggers, ...

The triggers are automatically created, the only thing needed are the annotations.

I've stated that before from time to time now. I am not happy with how the Update system works at the moment as well, BUT I think it's a reasonable enough default implementation for now.

It actually has some benefits when you think of it. In an application where a DB Admin doesn't allow for code to generate any database related code, the automatic generation from the info we find from the JPA provider might be problematic as 1. it's less obvious to the user what's happening under the hoods and 2. the information we can get from the JPA provider is actually quite limited. There is no way (as far as I can tell) that we can get information about what data-type the IDs are, so this might end up being a lot of guess-work.